### PR TITLE
User/lg/feature/ecmwf forecast

### PR DESCRIPTION
**Pull request by lawi119**
## Description

This PR provides the following functionality for the ecmwf-forecast dataset:
1. Split up the ecmwf-forecast collection into 7 different collections (ecmwf-forecast-wave-fc, ecmwf-forecast-cda-fc, ecmwf-forecast-oper-fc, ecmwf-forecast-enfo-ep, ecmwf-forecast-waef-ef, ecmwf-forecast-waef-ep, ecmwf-forecast-scwv-fc).
2. Update Dockerfile,requirements.txt to use the latest version of stactools-packages/ecmwf-forecast which creates kerchunk indices for the 7 collections. Note: Currently, this functionality only supports the wave-fc collection.
3. Modify .py scripts in the Kerchunk python package to write/read from the postgres database correctly.

Scripts created:
1. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-enfo-ep/description.md
2. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-enfo-ep/template.json
3. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-oper-fc/description.md
4. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-oper-fc/template.json
5. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-scda-fc/description.md
6. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-scda-fc/template.json
7. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-scwv-fc/description.md
8. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-scwv-fc/template.json
9. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-waef-ef/description.md
10. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-waef-ef/template.json
11. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-waef-ep/description.md
12. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-waef-ep/template.json
13. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-wave-fc/description.md
14. planetary-computer-tasks/datasets/ecmwf-forecast/collection/ecmwf-forecast-wave-fc/template.json
15. planetary-computer-tasks/datasets/ecmwf-forecast/modified_grib2.py: Modified to be able to write/read from postgres database correctly.
16. planetary-computer-tasks/datasets/ecmwf-forecast/modified_utils.py: Modified to be able to write/read from postgres database correctly.
17. planetary-computer-tasks/datasets/ecmwf-forecast/Kerchunk_test_notebook.ipynb: Created to test that the kerchunk indices for wave-fc STAC items are working properly.

Scripts changed:
1. planetary-computer-tasks/datasets/ecmwf-forecast/dataset.yaml: Updated to create 7 datasets, one for each collection. Updated to point to a docker image that has the latest version of stactools-packages/ecmwf-forecast.
2. planetary-computer-tasks/datasets/ecmwf-forecast/ecmwf_forecast.py: Updated to only create a STAC item for grib2 files that are not corrupt. Added in a section at the bottom to do two assertion checks.
3. planetary-computer-tasks/datasets/ecmwf-forecast/Dockerfile: Add two "COPY" lines to copy modified_utis.py and modified_grib2.py into the Kerchunk python package installed in the docker.
4. planetary-computer-tasks/datasets/ecmwf-forecast/requirements.txt: Add in kerchunk and ecmwflibs as necessary packages to install.
5. planetary-computer-tasks/datasets/ecmwf-forecast/README: Add the Ingest collection section. Update the Update workflow section.

Scripts removed:
1. planetary-computer-tasks/datasets/ecmwf-forecast/collection/template.json: No longer needed since the original collection was split into 7 collections.
2. planetary-computer-tasks/datasets/ecmwf-forecast/collection/description.md: No longer needed since the original collection was split into 7 collections.

Fixes # (issue)

N/A

## Type of change

- [Y] New feature (non-breaking change which adds functionality): The ECMWF dataset will process as 7 separate collection with kerchunk indices created for the wave-fc collection.
- [Y] Breaking change (fix or feature that would cause existing functionality to not work as expected): The ECMWF dataset will no longer process as one unified collection.
- [Y] This change requires a documentation update

## How Has This Been Tested?

Steps to test:
1. Git clone planetary-computer-tasks, set up, activate virtual environment in pctasks

. .venv/bin/activate
pctasks profile set test

2. Add dataset for first time

pctasks dataset ingest-collection --dataset ~/pr/planetary-computer-tasks/datasets/ecmwf-forecast/dataset.yaml --collection ecmwf-forecast-wave-fc

3. Process new items

pctasks dataset process-items -d ~/pr/planetary-computer-tasks/datasets/ecmwf-forecast/dataset.yaml --collection ecmwf-forecast-wave-fc kerchunk-lg --arg registry pccomponentstest.azurecr.io --upsert --submit

Note: You can change dataset.yaml to only process one day's worth of data by setting uri: blob://ai4edataeuwest/ecmwf/20231019/ for the ecmwf-forecast-wave-fc collection.

4. To test the kerchunk indices are working properly

Run through the notebook planetary-computer-tasks/datasets/ecmwf-forecast/Kerchunk_test_notebook.ipynb.

## Checklist:

Please delete options that are not relevant.

- [Y] I have performed a self-review
- [N] Changelog has been updated - not needed
- [Y] Documentation has been updated
- [N] Unit tests pass locally (./scripts/test) - not needed, added test to bottom of ecmwf_forecast.py instead
- [N] Code is linted and styled (./scripts/format) - not needed, added test to bottom of ecmwf_forecast.py instead

### CODE REVIEW

File: datasets/ecmwf-forecast/collection/ecmwf-forecast-enfo-ep/template.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    "item_assets": {
```
Comment (TomAugspurger) [inline datasets/ecmwf-forecast/collection/ecmwf-forecast-enfo-ep/template.json:75]: Just remembered, we'll want to update this to ensure that we get efficient compression on the STAC items in the database. Can chat more, but we want to include as much of `kerhcunk:indices` as possible here, everything that's static across all items in the dataset. Hopefully stuff like the `.zattrs` and `.zarray` are static, but maybe not.

File: datasets/ecmwf-forecast/requirements.txt [hunk 2]
```diff
@@ -1 +1,3 @@
-git+https://github.com/stactools-packages/ecmwf-forecast@1c9726e6a4f38e1c15648df6d45a5c8432ce92a9
\ No newline at end of file
+git+https://github.com/stactools-packages/ecmwf-forecast
```
Comment (TomAugspurger) [inline datasets/ecmwf-forecast/requirements.txt:1]: Let's pin this to a specific commit once https://github.com/stactools-packages/ecmwf-forecast/pull/5 is merged.

File: datasets/ecmwf-forecast/Dockerfile [hunk 3]
```diff
@@ -68,6 +68,8 @@ RUN cd /opt/src/pctasks/dataset && \
 COPY ./datasets/ecmwf-forecast/requirements.txt /opt/src/datasets/ecmwf-forecast/requirements.txt
 RUN python3 -m pip install -r /opt/src/datasets/ecmwf-forecast/requirements.txt
 
+COPY ./datasets/ecmwf-forecast/modified_utils.py /opt/conda/lib/python3.8/site-packages/kerchunk/utils.py
```
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply datasets/ecmwf-forecast/Dockerfile:71]: Looks like `modified_utils.py` can be achieved by something like

```python
for k, v in refs.items():
    if isinstance(v, str) and not v.startswith("base64:"):
        out[k] = (b"base64:" + base64.b64encode(v)).decode()
```

(untested)

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply datasets/ecmwf-forecast/Dockerfile:71]: Hopefully that'll work for grib2.py as well.

[the remark this answers is not in the document]
Comment (lawi119) [reply datasets/ecmwf-forecast/Dockerfile]: Added base64 encoding to stactools-packages/ecmwf-forecast instead. Removed the two modified files and COPY statement from Dockerfile. Done.

File: datasets/ecmwf-forecast/collection/ecmwf-forecast-enfo-ep/template.json (hunk 1 above)
Comment (lawi119) [reply datasets/ecmwf-forecast/collection/ecmwf-forecast-enfo-ep/template.json:75]: Done.

File: datasets/ecmwf-forecast/requirements.txt (hunk 2 above)
Comment (lawi119) [reply datasets/ecmwf-forecast/requirements.txt]: Done.

**Comment by TomAugspurger** [comment]
Merging this now. The `ecmwf-forecast-wave-fc` is completed. Some of the changes (like embedding the static kerchunk:indices in the Collection's `item_assets`) will need to be done for the remaining collections.

**Pull request merged**